Content.Server/StationEvents/EventManagerSystem.cs

```diff
@@ -44,12 +46,77 @@ public string RunRandomEvent()
         }
 
         var ent = GameTicker.AddGameRule(randomEvent);
-        var str = Loc.GetString("station-event-system-run-event",("eventName", ToPrettyString(ent)));
+        var str = Loc.GetString("station-event-system-run-event", ("eventName", ToPrettyString(ent)));
         _chat.SendAdminAlert(str);
         Log.Info(str);
         return str;
     }
 
+    /// <summary>
+    /// Randomly runs an event from provided EntitySpawnCollection.
+    /// </summary>
+    public string? RunRandomEvent(List<EntitySpawnEntry>? limitedEventsList)
+    {
+        if (limitedEventsList != null)
+        {
+            var availableEvents = AvailableEvents(); // handles the player counts and individual event restrictions
+
+            if (availableEvents.Count == 0)
+            {
+                Log.Warning("No events were available to run!");
+                return null;
+            }
+
+            var selectedEvents = EntitySpawnCollection.GetSpawns(limitedEventsList, _random); // storage function for game rules, smh my head.
+            // fuck it though, it works and gives us all the random selection utility we want.
+
+            Log.Info($"Picking from {limitedEventsList.Count} subsetted events");
+
+            var limitedEvents = new Dictionary<EntityPrototype, StationEventComponent>();
+            foreach (var eventid in selectedEvents)
+            {
+                if (!_prototype.TryIndex(eventid, out var eventproto))
+                {
+                    Log.Warning("An event ID has no prototype index!");
+                    continue;
+                }
+
+                if (eventproto.Abstract)
+                    continue;
+
+                if (!eventproto.TryGetComponent<StationEventComponent>(out var stationEvent, EntityManager.ComponentFactory))
+                    continue;
+
+                if (!availableEvents.ContainsKey(eventproto))
+                    continue;
+
+                limitedEvents.Add(eventproto, stationEvent);
+            }
+
+            var randomLimitedEvent = FindEvent(limitedEvents); // this picks the event, It might be better to use the GetSpawns to do it, but that will be a major rebalancing fuck.
+            if (randomLimitedEvent == null)
+            {
+                Log.Warning("The selected random event is null!");
+                return null;
+            }
+
+            if (!_prototype.TryIndex(randomLimitedEvent, out _))
+            {
+                Log.Warning("A requested event is not available!");
+                return null;
+            }
+
+            var ent = GameTicker.AddGameRule(randomLimitedEvent);
+            var str = Loc.GetString("station-event-system-run-event", ("eventName", ToPrettyString(ent)));
+            _chat.SendAdminAlert(str);
+            Log.Info(str);
+            return str;
+        }
+        Log.Warning("A scheduler has requested null as event!");
+        return null;
+    }
+
```

Comment:
Why does this even allow null inputs? Just make the method non-nullable instead of just logging a needless warning.

Reply:
Fixed. I changed the methods to void, since they were just returning a string, but as far as I can tell, nothing else interacts with that string anywhere.